python-sdk/tests/test_sdk.py

```diff
@@ -0,0 +1,261 @@
+    assert hasattr(rebuff_response, "max_heuristic_score")
+    assert hasattr(rebuff_response, "max_model_score")
+    assert hasattr(rebuff_response, "max_vector_score")
+    assert hasattr(rebuff_response, "injection_detected")
+
+
+def test_add_canary_word(rebuff: RebuffSdk, user_inputs: List[str]):
+    for user_input in user_inputs:
+        prompt = f"Tell me a joke about\n{user_input}"
+        buffed_prompt, canary_word = rebuff.add_canary_word(prompt)
+        assert canary_word in buffed_prompt
+
+
+@pytest.mark.parametrize(
+    "canary_word_leaked",
+    [True, False],
+    ids=["canary_word_leaked", "canary_word_not_leaked"],
+)
+def test_is_canary_word_leaked(
+    rebuff: RebuffSdk, user_inputs: List[str], canary_word_leaked: bool
+):
+    for user_input in user_inputs:
+        prompt = f"Tell me a joke about\n{user_input}"
+        _, canary_word = rebuff.add_canary_word(prompt)
+        log_outcome = False
+        if canary_word_leaked:
+            response_completion = (
+                f"<!-- {canary_word} -->\nTell me a joke about\n{user_input}"
+            )
+            leak_detected = rebuff.is_canary_word_leaked(
+                user_input, response_completion, canary_word, log_outcome
+            )
+            assert leak_detected is True
+
+        else:
+            response_completion = f"Tell me a joke about\n{user_inputs}"
+
+            leak_detected = rebuff.is_canary_word_leaked(
+                user_input, response_completion, canary_word, log_outcome
+            )
+            assert leak_detected is False
+
+
+def test_detect_injection_heuristics(
+    rebuff: RebuffSdk,
+    prompt_injection_inputs: List[str],
+    benign_inputs: List[str],
+    detect_injection_arguments: List[Union[float, bool]],
+):
+    max_heuristic_score = detect_injection_arguments[0]
+    max_vector_score = detect_injection_arguments[1]
+    max_model_score = detect_injection_arguments[2]
+    check_heuristic = True
+    check_vector = detect_injection_arguments[4]
+    check_llm = detect_injection_arguments[5]
+
+    for prompt_injection in prompt_injection_inputs:
+        rebuff_response = rebuff.detect_injection(
+            prompt_injection,
+            max_heuristic_score,
+            max_vector_score,
+            max_model_score,
+            check_heuristic,
+            check_vector,
+            check_llm,
+        )
+        assert rebuff_response.heuristic_score > max_heuristic_score
+        assert rebuff_response.injection_detected is True
+
+    for input in benign_inputs:
+        rebuff_response = rebuff.detect_injection(
+            input,
+            max_heuristic_score,
+            max_vector_score,
+            max_model_score,
+            check_heuristic,
+            check_vector,
+            check_llm,
+        )
+        assert rebuff_response.heuristic_score < max_heuristic_score
+        assert rebuff_response.injection_detected is False
```

Comment:
An approach you could take that would cut down on code:
```py
def test_detect_injection_heuristics(
    rebuff: RebuffSdk,
    prompt_injection_inputs: List[str],
    benign_inputs: List[str],
    detect_injection_arguments,
):
    detect_injection_arguments["check_heuristic"] = True

    user_inputs = prompt_injection_inputs + benign_inputs
    expect_detected = [True] * len(prompt_injection_inputs) + [False] * len(benign_inputs)
    for index, input in enumerate(user_inputs):
        rebuff_response = rebuff.detect_injection(
            input,
            **detect_injection_arguments,
        )
        assert (rebuff_response.heuristic_score > detect_injection_arguments["max_heuristic_score"]) == expect_detected[index]
        assert rebuff_response.injection_detected == expect_detected[index]
```
My code could be improved to make `user_inputs` and `expect_detected` fixtures.

Only accept my suggestion if you think it's better. It might be slightly harder to follow than the less compact approach you have.

Reply:
Thank you for suggesting this. I learnt new ways of making the code compact, appreciate you sharing it.  Though, also feel it is not as straight forward as the less-compact version. I think in the interest of code-readability, would prefer the less-compact version.